Guide/src/dev_guide/dev_tools/xflowey.md

```diff
@@ -10,13 +10,885 @@ for writing maintainable, cross-platform automation.
+working on OpenVMM automation.
+
+## Table of Contents
+
+1. [Core Concepts](#core-concepts)
+2. [Pipelines](#pipelines)
+3. [Artifacts](#artifacts)
+4. [Flowey Nodes](#flowey-nodes)
+5. [Variables: ReadVar and WriteVar](#variables-readvar-and-writevar)
+6. [Emitting Steps](#emitting-steps)
+7. [Runtime Services](#runtime-services)
+8. [Node Design Philosophy](#node-design-philosophy)
+9. [Common Patterns](#common-patterns)
+
+---
+
+## Core Concepts
+
+### Two-Phase Execution Model
+
+Flowey operates in two distinct phases:
+
+1. **Build-Time (Resolution Phase)**: When you run `cargo xflowey`, flowey
+   constructs a directed acyclic graph (DAG) of steps by:
+   - Instantiating all nodes
+   - Processing their requests
+   - Resolving dependencies between nodes via variables
+   - Determining the execution order
+
+2. **Runtime (Execution Phase)**: The generated flow is executed, and steps run
+   in the computed order. During runtime:
+   - Variables are read and written with actual values
+   - Commands are executed
+   - Artifacts are published/consumed
+   - Side effects occur
+
+This separation allows flowey to:
+- Validate the entire workflow before execution
+- Generate YAML for CI systems (ADO, GitHub Actions)
+- Optimize step ordering and parallelization
+- Catch dependency errors at build-time
+
+### Backend Abstraction
+
+Flowey supports multiple execution backends:
+
+- **Local**: Runs directly on your development machine via bash or direct
+  execution
+- **ADO (Azure DevOps)**: Generates ADO Pipeline YAML
+- **GitHub Actions**: Generates GitHub Actions workflow YAML
+
+**Important**: Nodes should be written to work across ALL backends whenever 
+possible. Relying on `ctx.backend()` to query the backend or manually emitting 
+backend-specific steps (via `emit_ado_step` or `emit_gh_step`) should be 
+avoided unless absolutely necessary. Most automation logic should be 
+backend-agnostic, using `emit_rust_step` for cross-platform Rust code that 
+works everywhere.
+
+---
+
+## Pipelines
+
+A **Pipeline** is the top-level construct that defines a complete automation
+workflow. Pipelines consist of one or more **Jobs**, each of which runs a set
+of **Nodes** to accomplish specific tasks.
+
+### Defining a Pipeline
+
+```rust
+fn into_pipeline(self, backend_hint: PipelineBackendHint) -> anyhow::Result<Pipeline> {
+    let mut pipeline = Pipeline::new();
+    
+    // Define a job that runs on Linux x86_64
+    let job = pipeline
+        .new_job(FlowPlatform::Linux(FlowPlatformLinuxDistro::Ubuntu), FlowArch::X86_64, "build")
+        .finish();
+    
+    Ok(pipeline)
+}
+```
+
+### Pipeline Jobs
+
+Each `PipelineJob` represents a unit of work that:
+- Runs on a specific platform and architecture
+- Can depend on artifacts from other jobs
+- Can be conditionally executed based on parameters
+- Emits a sequence of steps that accomplish the job's goals
+
+Jobs are configured using a builder pattern:
+
+```rust
```

Comment:
Suggest you move this to a doc comment in the code, and then you can just refer to that from the guide.